src/test/java/com/orbitz/consul/cache/ServiceCatalogCacheTest.java

```diff
@@ -0,0 +1,58 @@
+package com.orbitz.consul.cache;
+
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.orbitz.consul.BaseIntegrationTest;
+import com.orbitz.consul.model.catalog.CatalogService;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.*;
+
+public class ServiceCatalogCacheTest extends BaseIntegrationTest {
+    private static final Logger log = LoggerFactory.getLogger(ServiceCatalogCacheTest.class);
+
+    @Test
+    public void testWatchService() {
+        String name = UUID.randomUUID().toString();
+        String serviceId1 = createAutoDeregisterServiceId();
+        String serviceId2 = createAutoDeregisterServiceId();
+
+        List<Map<String, CatalogService>> result = new CopyOnWriteArrayList<>();
+        CountDownLatch finish = new CountDownLatch(1);
+
+        ServiceCatalogCache cache = ServiceCatalogCache.newCache(client.catalogClient(), name);
+        cache.addListener(serviceMap -> {
+            result.add(serviceMap);
+            if (result.size() == 3)
+                finish.countDown();
+        });
+
+        cache.start();
+
+        client.agentClient().register(20001, 20, name, serviceId1);
```

Comment:
if you do not insert a pause between this two lines, the test may become flaky

Reply:
fixed